### PR TITLE
Initial shot at structural equal and bug fixes in cpp_generator.py

### PR DESCRIPTION
**Pull request by electriclilies**
I added a method to do structural equality and changed the way relax SEqualReduce methods are generated and also some basic structural equality tests

The structural equality method only works on functions not relax modules right now, since relax modules are just dictionaries right now (are we going to change the relax modules to be an object? if not then I can just loop through that dictionary in the assert_structural_equal method).

### CODE REVIEW

File: objectgen/objectgen/object_def.py [hunk 1]
```diff
@@ -15,6 +15,7 @@ class ObjectField:
     field_name: str
     field_type: Type
     is_binding: bool = False
+    check_equality: bool = True
```
Comment (tkonolige) [inline objectgen/objectgen/object_def.py]: Would it make sense for this to be called `is_metadata`?

**Review by tkonolige — APPROVED** [review]
Looks mostly good.

File: objectgen/objectgen/object_def.py (hunk 1 above)
Comment (electriclilies) [reply objectgen/objectgen/object_def.py]: In the case of Spans, that makes sense, but I also had to set this false for function names. if function names can also count as metadata I think it works

Comment (tkonolige) [reply objectgen/objectgen/object_def.py]: arguably function names are meta data if you are only looking at structure

File: python/tvm/relax/frontend.py [hunk 2]
```diff
@@ -51,11 +52,6 @@ def __init__(self, definition_scope, diag_ctx):
         self.module = {}
         super().__init__()
 
-    def span_to_span(self, span):
```
Comment (jroesch) [inline python/tvm/relax/frontend.py:54]: I don't think this code is equivalent to tvm_span_from_synr, we need to be really careful about the string -> source name mapping.

File: objectgen/objectgen/object_def.py (hunk 1 above)
Comment (jroesch) [reply objectgen/objectgen/object_def.py]: I think we should keep it generic potentially like use_in_{equal/hash/etc} this is pretty much deriving for all of object gen so it shouldn't be specific to AST imo

**Pull request merged**